[2기-C 남득윤] JPA 게시판 미션 - 리포지토리 레이어, 서비스 레이어

ndy2 opened the pull request:
📌 과제 설명
회원과 게시글로 구성된 게시판 api구현 미션입니다.

👩‍💻 요구 사항과 구현 내용
과제 [링크](https://ancient-citron-06b.notion.site/SpringBoot-Part4-7-632dd10678164daca96a5ea185a02b60)입니다.

 API 구현 전 리포지토리와 서비스 레이어에 대한 PR입니다.

## ✅ 피드백 반영사항
- 사용자 검증 로직에서 불필요한 분기문을 제거하였습니다.
- 테스트 명에서 구체적인 숫자 제거하였습니다.
- 정적 팩토리 메서드를 제거하고 생성자를 사용하였습니다.
   - 바꾸지 않았었는데 읽고있는 리팩터링 책에서 코드 스멜로 소개하는 `추측성 일반화` 를 보고 너무 찔려서 바꾸었습니다...

## ✅ PR 포인트 & 궁금한 점
- `Post`
변경감지를 위한 Post 엔티티의 setter 메서드를 하나로 묶어 update라는 이름으로 잡고 자기자신을 리턴하게 했는데
이부분이 조금 이상한것 같기도 하네요...

- `PostRepository`
페이징 쿼리에서 `fetch join`를 사용하면 직접 count 쿼리를 작성해줘야 한다고 해서 조금 지저분한것 같아
찾아보니 `@EntityGraph` 라는 방식있어서 전체적으로 `@EntityGraph` 을 사용하는 것으로 통일 하였습니다.

- `AssertPost`
`post` 도메인에 대한 검증시 로직이 많이 중복되서 검증에 대한 정적 메서드를 제공하는 유틸성 클래스를 구현하였습니다.

- `PostServiceTest`, `PostRepositoryTest`
검증하고 싶은 로직 앞에서 영속성 컨텍스트를 초기화 해 주었는데 이 과정이 꼭 필요한 것인지 궁금합니다.

- `Post`, `PostTest`
작성자만이 게시글을 수정할 수 있다는 규칙을 자체적으로 추가하였는데
이부분에 대한 테스트 코드를 작성하는 것이 어려운데 어떤식으로 풀어야 하는지 궁금합니다.

  엔티티의 id 세팅에 jpa를 사용함에 따라 id 필드를 도메인 레벨의 테스트에서 활용하는 것이 제한되고 있습니다.
  fixture 를 제공할때 jpa를 활용해서 id 필드가 채워진 엔티티를 도메인 테스트에 제공하는 방식을 생각했는데 
  그러면 도메인 테스트가 스프링이나 jpa 에 의존하게 되어서 이상해지는거 같습니다..
  그렇다고 리플렉션을 쓰는것도 이상한거 같고 마땅한 방법이 있을까요?
  
  이에 현재 `Post.checkEditable` 메서드의 검증이 불가능한 상태입니다.
  해결된다면 이 함수를 `private` 으로 닫고 `udpate` 메서드로 옮기겠습니다.
  
  사실 마찬가지로 `Post` 엔티티의 생성시에도 전달받은 `User` 인스턴스의 id 값에 대한 
  `null` check 도 해주고 싶은데 이부분도 같은 이유로 생략하였습니다.

  해결된다면 추가하겠습니다.

ndy2 committed:
refactor: 코드 리뷰 반영

테스트 메서드 이름에 구체적인 숫자 제거
검증 로직 분기문 제거

ndy2 committed:
feat: 커스텀 루트 예외 - BoardRuntimeException 구현

ndy2 committed:
feat: PostRepository 구현

ToDo - JpaPostRepository 로 추상화

ndy2 committed:
feat: PostService 조회성 메서드 구현

findAll, findById

ndy2 committed:
test: PostServiceTest 조회성 메서드 테스트

ndy2 committed:
test: PostRepositoryTest 조회성 메서드 테스트

ndy2 committed:
feat: PostService 작성, 수정 메서드 구현

ndy2 committed:
test: PostServiceTest 작성, 수정 메서드 테스트

ndy2 committed:
refactor: post,writer 검증 메서드 추출

PostTest.assertPost, PostTest.assertWriter

ndy2 committed:
rename: 리포지토리들 repository 패키지로 이동

ndy2 committed:
rename: 패키지 이름 변경, 이동

리포지토리들 repository 패키지로 이동
auditing -> base

ndy2 committed:
feat: 게시글 전체 조회 Paging 구현

ndy2 committed:
test: 게시글 전체 조회 Paging 구현

ndy2 committed:
refactor: PostRepository 어노테이션 변경

@Query -> @EntityGraph

ndy2 committed:
test: 검증 메서드 추출

assertPost,assertWriter -> assertPostWithWriter

ndy2 committed:
test: id로 조회 검증 방식 통일

테스트 메서드에서 삽입후 조회 -> setUp 데이터 조회로 변경

ndy2 committed:
refactor: code reformatting

ndy2 committed:
test: post 검증 메서드 클래스로 분리

PostTest -> AssertPost

ndy2 committed:
refactor: BoardRuntimeException 추상 클래스로 변경

ndy2 committed:
refactor: Post.create 제거 -> 생성자 사용

ndy2 committed:
refactor: 수정시 작성자 검증 로직 위치 변경

PostService.checkEditable -> Post.checkEditable

ndy2 committed:
rename: Fixture prefix create -> sample

ndy2 committed:
feat: Post 수정 validation 추가

ndy2 committed:
test: Post 수정 테스트

ndy2 committed:
refactor: User.create 제거 -> 생성자 사용

ndy2 committed:
refactor: checkEditable 메서드 시그니처 변경

checkEditable(Long, User) -> checkUpdatable(User)

woogiekim commented on src/main/java/com/study/board/domain/post/domain/Post.java:
```diff
@@ -48,8 +48,31 @@ private Post(String title, String content, User writer) {
+        }
     }
 
+    private boolean hasOfWriter(User writer) {
```
이 기능에 hasOfWriter 라는 이름은 어색한거같은데요

isWriter 가 더 괜찮지 않나요?

ndy2 replied:
저는 `hasOfWriter` 혹은 `hasWriter` 가 더 와닿는거 같습니다...
게시글의 입장에서 사용자를 has 하는거 같아서...

`post.isWriter(writer)` 하면 개인적으로 이상한것 같아요..

추가로 생각해보니 `isWrittenBy` 도 좋은것 같습니다.

woogiekim commented on src/main/java/com/study/board/domain/post/domain/Post.java:
```diff
@@ -48,8 +48,31 @@ private Post(String title, String content, User writer) {
+        checkArgument(hasText(content), "content - 글자를 가져야함");
+    }
+
+    public void checkEditable(Long postId, User editor) {
+        if (!hasOfWriter(editor)) {
+            throw new PostEditAccessDeniedException(postId);
+        }
     }
```
서비스에서 하지말고 도메인에서 update 할 때 하는건 어떤가요?

ndy2 replied:
이부분에 대해서 테스트에 제약이 생겨 일단 이런식으로 진행하였습니다.

테스트에 아이디 필드의 세팅이 필수 적인데
이부분이 jpa 에 심하게 의존함으로써 도메인 테스트 자체가 jpa 에 의존하는 문제가 있습니다.

이에 관해 궁금한 점을 위 코멘트에 작성하였는데 확인 부탁 드립니다!

woogiekim commented on src/main/java/com/study/board/domain/post/repository/PostRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+public interface PostRepository extends JpaRepository<Post, Long> {
+
+    @EntityGraph(attributePaths = "writer")
```
이걸 하는 이유는 뭔가요?

ndy2 replied:
게시글을 조회하며 join 문으로 바로 작성자 정보를 조회 하기 위해 사용하였습니다.

fetch join + 페이징 쿼리를 사용할때 data jpa 가 자동으로 count query 를 만들어 주는 기능에 문제가 생겨
직접 count query 를 명시해줘야 한다고 합니다. 이는 좀 지저분 한 것 같아서
찾아보니 `@EntityGraph` 로 fetch 를 걸면 된다고 해서 적용하였습니다.

다른 방법이 있을까요?

참고한 링크입니다. [페이징 질문 - 인프런](https://www.inflearn.com/questions/167730)

woogiekim commented on src/main/java/com/study/board/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,57 @@
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class PostService {
```
인터페이스를 만들지 않는 이유가 있나요?

ndy2 replied:
혼자 진행할 때는 너무 복잡해 지는 느낌이라 다형성을 활용할 때만 인터페이스를 만드는 것을 좋아했는데

코드리뷰를 받고 함께 프로젝트를 진행할 때에는 특히 서비스 클래스에 대해서는
인터페이스를 통해 사용가능한 메서드를 한눈에 보여 주는것도 좋을것 같네요.

woogiekim commented on src/test/java/com/study/board/domain/post/AssertPost.java:
```diff
@@ -0,0 +1,34 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class AssertPost {
```
이게 과연 얼마나 재사용할 수 있을까요?

ndy2 replied:
재사용이라는 것이 변화에 대응하는 그런 의미 맞을까요?
말씀하시는 재사용성의 의미가 잘 이해되지 않습니다...

일단 저는 재사용성에 대한 고려 없이 검증 코드가 너무 중복되는 것 같아서 유틸성 클래스로 뺏습니다.

ndy2 replied:
아 이것 계속 고민하다가 멘토님 코드를 참고했는데

entity 의 equality 를 id baseId 의 EqualsAndHashcode 오버라이드로 검증할 수 있군요.
진짜 완전 멋지네요

id만 base 로 빼는게 여기서 또 이렇게 활용되네요... 와 .... 완전 멋지네요 정말

woogiekim commented on src/test/java/com/study/board/domain/post/service/PostServiceTest.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Test
+    void 없는_게시글_수정_하면_실패() {
+        Long illegalPostId = -1L;
+
+        assertThatThrownBy(() -> postService.edit(illegalPostId, "수정제목", "수정내용", writer.getId()))
+                .isInstanceOf(PostNotFoundException.class);
+    }
```
작성자가 글 수정을 한다 -> 없는 글이다

라는 상황이 정상적인 상황일까요?

ndy2 replied:
아래 예외발생 코멘트 시리즈에도 마찬가지로
 
앞단 (필터레벨) 에서 잡아야 하는 예외 상황을 굳이 서비스 까지 끌고와서 구현하다보니 생긴
이상한 상황인것 같습니다.

요구사항에 따른 구현 위치를 명확히하고
TODO 로 분리해서 이런 불필요한 개발을 하지 않도록 해야겠네요..

ndy2 replied:
게시글 및 작성자의 유효성 (존재)에 대한 로직은 앞단 filter 레벨에서 추후에 구현 하는 것으로 가정하고
`Post.write`, `Post.edit` 메서드에서는 전달 받은 id 값들이 모두 유효하다는 가정을 하고 리팩토링 하겠습니다.

ndy2 replied:
그래도 어짜피 조회 하면 optional 로 반환이 되기 때문에 앞단에서 필터링 되었음이 확실하다고 해도
그냥 .get() 을 호출하고 테스트를 생략하는 것은 이상한것 같아서 BoardNotFoundException 이라는 하나의 예외로 모두 처리 하도록 하였습니다.

컨트롤러를 개발한다면 이 예외는 따로 더 신경써서 처리해야 겟네요..

woogiekim replied:
해당사항은 버그거나 공격이라 get 으로 하시거나 단순 예외를 던지면 되지 않나 싶어요

woogiekim commented on src/test/java/com/study/board/domain/post/service/PostServiceTest.java:
```diff
@@ -0,0 +1,151 @@
+
+        assertThatThrownBy(() -> postService.edit(writtenPostId1, "수정제목", "수정내용", illegalUserId))
+                .isInstanceOf(UserNotFoundException.class);
+    }
```
애초에 없는 사용자가 게시글을 수정하려고 한다는거 자체가 말이 안되지 않을까요?

커스텀 예외로 지정해서 사용자에게 메시지를 보내줄건가요?

woogiekim commented on src/test/java/com/study/board/domain/post/service/PostServiceTest.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Test
+    void 남의_게시글_수정하려하면_실패() {
+        Long anotherUserId = userRepository.save(new User("다른 사용자", null)).getId();
+
+        assertThatThrownBy(() -> postService.edit(writtenPostId1, "수정제목", "수정내용", anotherUserId))
+                .isInstanceOf(PostEditAccessDeniedException.class);
```
남의 글을 수정하려고 하는거 자체가 사실 공격이라고 봐도 되지 않을까요?

메시지를 보내줘야할까요?

ndy2 replied:
음.. 그냥 access denied 정도의 응답이 적당하다는 의견일까요?

그러면 PostEditAccessDeniedException을 ControllerAdvice 에서 잡아서 
AccessDenied 정도의 레벨로 변경해서 응답하는게 필요할까요?

woogiekim commented on src/test/java/com/study/board/fixture/Fixture.java:
```diff
@@ -1,13 +1,15 @@
 
-    public static User createUser() {
-        return User.create("득윤", null);
+    public static User sampleUser() {
```
sample 이라는네이밍이 별로인거같습니다...

create나 a 이런게 좋아보입니다.

ndy2 replied:
앗.. 그런가요 ㅜ.ㅜ
읽고 있는 책에서 sample 이라는 네이밍으로 fixture 를 소개하길래 꽤 통용되는 네이밍인줄 알았는데..

create 도 좋은거 같습니다. create 으로 다시 돌리겠습니다.

woogiekim commented on src/test/java/com/study/board/domain/post/service/PostServiceTest.java:
```diff
@@ -0,0 +1,151 @@
+    @Autowired
+    UserRepository userRepository;
+
+    @Autowired
+    EntityManager em;
```
이거는 왜 필요한건지 잘 모르겠습니다
왜 필요한가요?

ndy2 replied:
추가한 데이터에 대한 영속성 컨텍스트를 비워주기 위해 주입받았습니다.

이부분이 저도 고민되어서 질문 게시판에 질문을 남겨 홍구님께 답변도 들었는데 멘토님 의견이 궁금합니다.
홍구님은 비워준다고 하셨습니다. 그 방식에 대한 자세한 내용을 듣지 못하였습니다.

질문 내용
> JPA 를 활용한 리포지토리, 서비스 테스트 작성시 테스트의 하고자 하는 메서드 호출 전에 영속성 컨텍스트를 비워주는 과정이 필요하다고 생각했는데 이부분에 대한 의견이 궁금합니다.
데이터를 1차 캐시에서 가져오면 테스트의 성격이나 의도에 부합한다고 볼 수 없다는 입장에서 em.clear 의 호출이 필요하다고 생각했습니다.
하지만 서비스의 테스트가 리포지토리 계층의 특정 기술에 해당 되는 jpa의 엔티티 매니저를 주입 받아 사용하는 것이 이상한것 같기도 합니다.

woogiekim replied:
save하는 경우는 1차 캐시를 비우고

조회 쿼리가 날라가게 의도할수있을거같긴합니다

이런식의 경우가 아니면 굳이 그래야하나 싶어요

ndy2 replied:
테스트 대상 메서드가 save인 경우를 말씀하시는 거 군요

음.. 그러면 리포지토리 테스트에서는 1차 캐시를 비우고
서비스 테스트에서는 1차 캐시나 영속성 컨텍스트 관련 맥락과 이슈는
 리포지토리에서 처리 했다고 가정하고 그 구문을 지우는 것으로 하고

리포지토리 테스트에서는 save 호출 이후 1차 캐시를 비워서 쿼리를 확인 하도록 하겠습니다.
아무래도 처음 공부하는 입장이다 보니 쿼리를 눈으로 확인 하지 않으면 불안하네요 ㅜ.ㅜ

woogiekim reviewed (changes requested):
고생하셨습니다!

코멘트 보시고 확인해주세요 :)

ndy2 committed:
refactor: 메서드 이름 변경

Post.hasOfWriter -> Post.isWrittenBy

ndy2 committed:
refactor: PostService 인터페이스 추가

ndy2 committed:
refactor: change prefix of fixture methods

sample -> create

ndy2 committed:
refactor: 여러 예외 상황에 대한 에러 통일

PostEditAccessDeniedException
PostNotFoundException
UserNotFoundException
-> BoardNotFoundException

ndy2 committed:
move: 예외 패키지 root 로 이동

ndy2 committed:
rename: baseEntity -> BaseIdEntity

ndy2 committed:
refactor: BaseIdEntity EqualsAndHashcode

ndy2 committed:
refactor: 리뷰 기반 개선

검증 방식 AssertPost -> eqauls 기반, 필요시 필드 검증
커스텀 예외 제거 -> IllegalArgumentException 사용

ndy2 commented:
리뷰 바탕으로 리팩토링진행 하였습니다.
---
1. 테스트시 엔티티 검증 방식을 변경하였습니다.
  - 기존에는 모든 상황에서 필드를 검증 하였고 이를 AssertPost 라는 유틸성 클래스를 통해 처리 하였습니다. 이를 아래의 두 상황에 대하여 검증방식을 구분하는 것으로 변경하였습니다.
  -  일반적인 조회 상황 : `BaseEntity` -> `BaseIdEntity`에 `@EqualsAndHashCode` 를 재정의 하여 Id 가 같으면 동일한 Entity 라는 기본 전제를 바탕으로 `eqauls` 로 검증
  -  생성, 수정에 관한 테스트 : 에서는 관심사가 필드의 세팅이라고 생각해서 대상 필드를 검증하였습니다. 


---
2. 정상적이지 않은 상황에 대한 예외처리 방식 변경
>   작성자가 글 수정을 한다 -> 없는 글이다 라는 상황이 정상적인 상황일까요?

혹시 위 코멘트가 예외를 커스텀으로 처리하는 것이 이상하다는  내용으로 이해했는데  맞을까요?

이해를 바탕으로 커스텀 예외를 제거하고 단순예외 (`IllegalArgumentException`)으로 처리하였습니다.


---
3. 게시글의 수정 전 체크 방식 변경 (`Post.checkUpdatable` -> `PostService.checkEditable`)
일단 update 라는 이름을 모두 edit 으로 변경하였습니다.
그리고 저 체크로직을 service 가 아닌 domain 으로 이동하는 것을 추천하셨는데 저는 이를
`Post.edit` :
``` java
public Post edit(String title, String content) {
        checkEditable();
        ...
       
       
    }
``` 
   이런식으로 `edit` 메서드 안으로 이동하라는 내용으로 이해했습니다.

  저도 이게 구조적으로 더 맞는 것 같은데 문제는 이를 테스트 하는 방식을 잘 모르겠습니다.
  `checkEditable` 메서드는 `writer` 필드의 id에 대한 참조를 통해 이루어 지는데 이값이 jpa 에 의해 세팅이 되어
  도메인 로직 만으로 세팅이 불가능 하고 이에 따라 테스트 하는데 제한이 있습니다.

그래서 일단 가장 문제가 되는 부분인 `isWrittenBy` 메서드 (id를 참조함) 만 도메인에서 제공하고
아예 `checkEditable` 메서드는 서비스로 빼버렸습니다.

이를 테스트하는 방법만 해결되면 위와 같이 `edit` 메서드 안으로 `checkEditable` 을 옮겨버리겠습니다.

ndy2 committed:
refactor: fix typo

woogiekim reviewed (approved):
리뷰토대로 조사을 많이 하신거 같습니다!
고생하셨습니다!
